packages/phpstan-rules/src/Rules/ClassNamespaceGuardRule.php

```diff
@@ -0,0 +1,168 @@
+    {
+        $classLikeName = $this->simpleNameResolver->getName($node);
+        if ($classLikeName === null) {
+            return [];
+        }
+
+        $namespace = $scope->getNamespace();
+        if ($namespace === null) {
+            return [];
+        }
+
+        foreach ($this->guards as $type => $allowedNamespacePatterns) {
+            $classReflection = $this->reflectionProvider->getClass($classLikeName);
+
+            if (! $classReflection->isSubclassOf($type)) {
+                continue;
+            }
+
+            $isInAllowedNamespace = false;
+            foreach ($allowedNamespacePatterns as $allowedNamespacePattern) {
+                if ($this->isClassLikeNameMatchedAgainstPattern($classLikeName, $allowedNamespacePattern)) {
+                    $isInAllowedNamespace = true;
+                    break;
+                }
+            }
+
+            if (! $isInAllowedNamespace) {
+                $errorMessage = sprintf(
+                    self::ERROR_MESSAGE,
+                    $classLikeName,
+                    json_encode($allowedNamespacePatterns, JSON_THROW_ON_ERROR),
+                    $classReflection->getNativeReflection()
+                        ->getNamespaceName(),
+                );
+
+                return [$errorMessage];
+            }
+        }
+
+        return [];
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition('Define in which namespaces (using *, ** or ? glob-like pattern matching) can classes extending specified class or implementing specified interface exist', [
+            new ConfiguredCodeSample(
+                <<<'CODE_SAMPLE'
+namespace App;
+
+// AbstractType implements \Symfony\Component\Form\FormTypeInterface
+use Symfony\Component\Form\AbstractType;
+
+class UserForm extends AbstractType
+{
+}
+CODE_SAMPLE
+                ,
+                <<<'CODE_SAMPLE'
+namespace App\Form;
+
+use Symfony\Component\Form\AbstractType;
+
+class UserForm extends AbstractType
+{
+}
+CODE_SAMPLE
+                ,
+                [
+                    'guards' => [
+                        'Symfony\Component\Form\FormTypeInterface' => ['App\Form\**'],
+                    ],
+                ]
+            ),
+            new ConfiguredCodeSample(
+                <<<'CODE_SAMPLE'
+namespace App;
+
+use Exception;
+
+class ProductNotFoundException extends Exception
+{
+}
+CODE_SAMPLE
+                ,
+                <<<'CODE_SAMPLE'
+namespace App\Exception;
+
+use Exception;
+
+class ProductNotFoundException extends Exception
+{
+}
+CODE_SAMPLE
+                ,
+                [
+                    'guards' => [
+                        Throwable::class => ['App\Exception\**', 'App\Services\**'],
+                    ],
+                ]
+            ),
+        ]);
+    }
+
+    private function isClassLikeNameMatchedAgainstPattern(string $className, string $namespaceWildcardPattern): bool
```

Comment:
This might be useful to reuse and cover with tests.
Could you extract it to service?